arch/risc-v/src/mpfs/mpfs_fpga_canfd.c

```diff
@@ -0,0 +1,2917 @@
+                                       MPFS_SYSREG_SUBBLK_CLOCK_CR_OFFSET)
+
+#define CANWORK                 HPWORK
+
+#define MPFS_CANFD_ID           0xCAFD
+
+/* For allocating the tx and rx CAN-FD frame buffer */
+
+#define POOL_SIZE               1
+#define TIMESTAMP_SIZE          sizeof(struct timeval)  /* To support
+                                                         * timestamping frame */
+
+/* For bit timing calculation */
+
+#define BT_COMPUTE_MAX_ERROR      50 /* 1/10th of % */
+#define BT_COMPUTE_SYNC_SEG       1
+
+#ifdef CONFIG_NETDEV_CAN_FILTER_IOCTL
+/* CAN hw filter support */
+
+#define HW_FILTER_A             0
+#define HW_FILTER_B             1
+#define HW_FILTER_C             2
+#define HW_FILTER_RANGE         3
+
+#define CAN_STD_ID              0
+#define CAN_EXT_ID              1
+#endif /* CONFIG_NETDEV_CAN_FILTER_IOCTL */
+
+/* Special address description flags for the CAN_ID */
+
+#define CAN_EFF_FLAG            0x80000000  /* EFF/SFF is set in the MSB */
+#define CAN_RTR_FLAG            0x40000000  /* remote transmission request */
+#define CAN_ERR_FLAG            0x20000000  /* error message frame */
+
+/* Valid bits in CAN ID for frame formats */
+
+#define CAN_SFF_MASK            0x000007FF /* standard frame format (SFF) */
+#define CAN_EFF_MASK            0x1FFFFFFF /* extended frame format (EFF) */
+#define CAN_ERR_MASK            0x1FFFFFFF /* omit EFF, RTR, ERR flags */
+
+/* CAN control mode */
+
+#define CAN_CTRLMODE_LOOPBACK       0x01  /* Loopback mode */
+#define CAN_CTRLMODE_LISTENONLY     0x02  /* Listen-only mode */
+#define CAN_CTRLMODE_3_SAMPLES      0x04  /* Triple sampling mode */
+#define CAN_CTRLMODE_ONE_SHOT       0x08  /* One-Shot mode */
+#define CAN_CTRLMODE_BERR_REPORTING 0x10  /* Bus-error reporting */
+#define CAN_CTRLMODE_FD             0x20  /* CAN FD mode */
+#define CAN_CTRLMODE_PRESUME_ACK    0x40  /* Ignore missing CAN ACKs */
+#define CAN_CTRLMODE_FD_NON_ISO     0x80  /* CAN FD in non-ISO mode */
+#define CAN_CTRLMODE_CC_LEN8_DLC    0x100 /* Classic CAN DLC option */
+#define CAN_CTRLMODE_TDC_AUTO       0x200 /* CAN transiver automatically
+                                           * calculates TDCV */
+#define CAN_CTRLMODE_TDC_MANUAL     0x400 /* TDCV is manually set up by user */
+
+/* TXT buffer */
+
+enum mpfs_can_txb_status
+{
+  TXT_NOT_EXIST       = 0x0,
+  TXT_RDY             = 0x1,
+  TXT_TRAN            = 0x2,
+  TXT_ABTP            = 0x3,
+  TXT_TOK             = 0x4,
+  TXT_ERR             = 0x6,
+  TXT_ABT             = 0x7,
+  TXT_ETY             = 0x8,
+};
+
+enum mpfs_can_txb_command
+{
+  TXT_CMD_SET_EMPTY   = 0x01,
+  TXT_CMD_SET_READY   = 0x02,
+  TXT_CMD_SET_ABORT   = 0x04
+};
+
+#define candebug _none
+
+/****************************************************************************
+ * Utility definitions
+ ****************************************************************************/
+
+#ifndef min
+#define min(a, b)  (a) < (b) ? a : b
```

Comment:
I still suggest adding parentheses around a and b after the ternary operation:

((a) < (b) ? (a) : (b))

Reply:
yes, fixed